### PR TITLE
Кальянов --Лабораторная работа #3

### PR DESCRIPTION
**Pull request by KalyanovDO**
@AndrewSmrnv, @VladimirTyagunov, please, take a look.

### CODE REVIEW

File: code/dijkstra-algorithm/Infrastructure/src/main/java/ru/unn/agile/dijkstraalgorithm/infrastructure/TextLogger.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+        writer = bufferedLogWriter;
+    }
+
+    @Override
+    public void log(final String s) {
+        try {
+            writer.write(now() + " > " + s);
```
Comment (vtyagunov) [inline code/dijkstra-algorithm/Infrastructure/src/main/java/ru/unn/agile/dijkstraalgorithm/infrastructure/TextLogger.java]: Move `>` separator to private static final field

File: code/dijkstra-algorithm/ViewModel/src/main/java/ru/unn/agile/dijkstraalgorithm/viewmodel/ViewModel.java [hunk 2]
```diff
@@ -106,56 +133,129 @@ private void updateVertexList() {
...
         return vertexList;
     }
+    public StringProperty logsProperty() {
+        return logs;
+    }
+    public final String getLogs() {
+        return logs.get();
+    }
+    public final List<String> getLogList() {
+        return logger.getLog();
```
Comment (vtyagunov) [inline code/dijkstra-algorithm/ViewModel/src/main/java/ru/unn/agile/dijkstraalgorithm/viewmodel/ViewModel.java]: If your class are initialized via empty constructor, you will get NullPointerException by calling this method.
So it's safer to check `logger` for null first, and then call `getLog` method

File: code/dijkstra-algorithm/ViewModel/src/main/java/ru/unn/agile/dijkstraalgorithm/viewmodel/ViewModel.java [hunk 3]
```diff
@@ -106,56 +133,129 @@ private void updateVertexList() {
...
+
+    private void logForAddedEdge() {
+        StringBuilder message = new StringBuilder(LogMessages.ADD_EDGE_WAS_PRESSED);
+        message.append("[")
+                .append(vertex1.get())
+                .append(",")
+                .append(vertex2.get())
+                .append(",")
+                .append(weight.get())
+                .append("]");
```
Comment (vtyagunov) [inline code/dijkstra-algorithm/ViewModel/src/main/java/ru/unn/agile/dijkstraalgorithm/viewmodel/ViewModel.java]: `logForAddedEdge` and `onExpressionTextFieldFocusChanged` have the same message construction
Its a good practice to use String.format() or even extract this as a method

**Review by vtyagunov — CHANGES_REQUESTED** [review]

File: code/dijkstra-algorithm/Infrastructure/src/main/java/ru/unn/agile/dijkstraalgorithm/infrastructure/TextLogger.java (hunk 1 above)
Comment (AndrewSmrnv) [reply code/dijkstra-algorithm/Infrastructure/src/main/java/ru/unn/agile/dijkstraalgorithm/infrastructure/TextLogger.java]: What for?

Comment (KalyanovDO) [reply code/dijkstra-algorithm/Infrastructure/src/main/java/ru/unn/agile/dijkstraalgorithm/infrastructure/TextLogger.java]: Done

File: code/dijkstra-algorithm/ViewModel/src/main/java/ru/unn/agile/dijkstraalgorithm/viewmodel/ViewModel.java (hunk 2 above)
Comment (KalyanovDO) [reply code/dijkstra-algorithm/ViewModel/src/main/java/ru/unn/agile/dijkstraalgorithm/viewmodel/ViewModel.java]: Done

File: code/dijkstra-algorithm/ViewModel/src/main/java/ru/unn/agile/dijkstraalgorithm/viewmodel/ViewModel.java (hunk 3 above)
Comment (KalyanovDO) [reply code/dijkstra-algorithm/ViewModel/src/main/java/ru/unn/agile/dijkstraalgorithm/viewmodel/ViewModel.java]: Done

**Comment by AndrewSmrnv** [comment]
👍

**Comment by vtyagunov** [comment]
👍

File: code/dijkstra-algorithm/Infrastructure/src/main/java/ru/unn/agile/dijkstraalgorithm/infrastructure/TextLogger.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.List;
+import java.util.Locale;
+
+public class TextLogger implements ILogger {
+    private final String filename;
+    private final BufferedWriter writer;
+    private static final String DATE_FORMAT_POINT = "yyyy-MM-dd HH:mm:ss";
```
Comment (VitKonevskiy) [inline code/dijkstra-algorithm/Infrastructure/src/main/java/ru/unn/agile/dijkstraalgorithm/infrastructure/TextLogger.java]: yyyy-MM-dd HH:mm:ss.SSS should be more usefull

File: code/dijkstra-algorithm/ViewModel/src/main/java/ru/unn/agile/dijkstraalgorithm/viewmodel/ViewModel.java [hunk 5]
```diff
@@ -85,8 +87,29 @@ public boolean isWeightInputCorrect() {
...
+            }
+        };
+        addingNewEdgeDisabled.bind(canCalculateBoolBinding.not());
+    }
+
     public void createGraph() {
+        StringBuilder message = new StringBuilder(LogMessages.CREATE_GRAPH_WAS_PRESSED);
         if (edgeList.isEmpty()) {
+            message.append(" unsuccessfully. List of edges is empty");
+            logger.log(message.toString());
```
Comment (VitKonevskiy) [inline code/dijkstra-algorithm/ViewModel/src/main/java/ru/unn/agile/dijkstraalgorithm/viewmodel/ViewModel.java]: What should happen if logger is null?

**Review by VitKonevskiy — CHANGES_REQUESTED** [review]

File: code/dijkstra-algorithm/Infrastructure/src/main/java/ru/unn/agile/dijkstraalgorithm/infrastructure/TextLogger.java (hunk 4 above)
Comment (KalyanovDO) [reply code/dijkstra-algorithm/Infrastructure/src/main/java/ru/unn/agile/dijkstraalgorithm/infrastructure/TextLogger.java]: It is!

File: code/dijkstra-algorithm/ViewModel/src/main/java/ru/unn/agile/dijkstraalgorithm/viewmodel/ViewModel.java (hunk 5 above)
Comment (KalyanovDO) [reply code/dijkstra-algorithm/ViewModel/src/main/java/ru/unn/agile/dijkstraalgorithm/viewmodel/ViewModel.java]: Checks for NPE were added

**Review by VitKonevskiy — APPROVED** [review]

**Pull request merged**